### PR TITLE
Added support for reading MultiLineString WKTs

### PR DESCRIPTION
**Pull request by jacobrec**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added support for reading MULTILINESTRING geometry in WKT format using function readWKTLineString.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit a80c7c080cdbb3bec4662501686133298b7a4a2d with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67647/a80c7c080cdbb3bec4662501686133298b7a4a2d/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67647/a80c7c080cdbb3bec4662501686133298b7a4a2d/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: docs/en/sql-reference/functions/geo/polygon.md [hunk 1]
```diff
@@ -170,6 +176,34 @@ SELECT readWKTLineString('LINESTRING (1 1, 2 2, 3 3, 1 1)');
...
+- `wkt_string`: The input WKT string representing a MultiLineString geometry.
+
+### Returned value
+
+The function returns a ClickHouse internal representation of the multilinestring geometry.
+
+### Example
+
+```sql
+SELECT wkt(readWKTMultiLineString('MULTILINESTRING ((1 1, 2 2, 3 3), (4 4, 5 5, 6 6))'));
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/geo/polygon.md:200]: I think the example should be just `readWKTMultiLineString('MULTILINESTRING ((1 1, 2 2, 3 3), (4 4, 5 5, 6 6))')` (without `wkt`)

File: src/Functions/geometryConverters.h [hunk 2]
```diff
@@ -411,6 +470,11 @@ static void callOnGeometryDataType(DataTypePtr type, F && f)
     else if (factory.get("LineString")->equals(*type) && type->getCustomName() && type->getCustomName()->getName() == "LineString")
         return f(ConverterType<ColumnToLineStringsConverter<Point>>());
 
+    /// We should take the name into consideration to avoid ambiguity.
+    /// Because for example both MultiLineString and Polygon are resolved to Array(Tuple(Point)).
```
Comment (alexey-milovidov) [inline src/Functions/geometryConverters.h:474]: Did you mean `Array(Array(Point))`?

File: tests/queries/0_stateless/03215_multilinestring_geometry.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
+-- { echoOn }
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/03215_multilinestring_geometry.sql:1]: Need at least one query with non-constant input.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/03215_multilinestring_geometry.sql:1]: (better with different array sizes of both the number of lines and the number of points)

**Comment by alexey-milovidov** [comment]
Most looks good to me,
also, I'd like it if @nikitamikhaylov would check the code around disambiguation data types based on special names (it looks ok, but maybe there are some considerations).

**Comment by nikitamikhaylov** [comment]
Looks good to me, all the comments are valid. Let's add more tests with non-constant input: for example create a table with a column of type `Array(Array(Point))`, call `wkt` from that column and make sure the result is serialized as `POLYGON`s - to ensure backward compatibility for non-qualified types.

**Comment by jacobrec** [comment]
Thanks for the feedback @alexey-milovidov and @nikitamikhaylov, I've added some additional tests and updated the comments/docs

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@jacobrec, Amazing, thank you!

**Comment by clickgapai** [comment]
Hi @jacobrec @alexey-milovidov — while reviewing this PR I found the following:

- **Bug (P2):** `readWKTMultiLineString` documentation example uses wrong function name `readWKTLineString` — https://github.com/ClickHouse/ClickHouse/issues/104974

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**